pkg/logql/log/parser_hints.go

```diff
@@ -57,30 +64,45 @@ func (p *parserHint) ShouldExtractPrefix(prefix string) bool {
 	return false
 }
 
-func (p *parserHint) NoLabels() bool {
-	return p.noLabels
+func (p *Hints) NoLabels() bool {
+	return p.noLabels || p.AllRequiredExtracted()
+}
+
+func (p *Hints) RecordExtracted() {
+	p.extracted++
+}
+
+func (p *Hints) AllRequiredExtracted() bool {
+	if len(p.requiredLabels) == 0 {
+		return false
+	}
+	return p.extracted == len(p.requiredLabels)
```

Comment:
What if a label is twice in a JSON document?

Reply:
This is actually really interesting. I thought duplicate json keys were invalid json and it turns out that _might_ not be the case. It's ambiguous enough that this is a good thing to cover.

https://stackoverflow.com/a/23195243

Reply:
lol, @jeschkies , this actually invalidates this whole approach. If duplicates can exist in the log -- as they can for `json` and `logfmt` -- there's no way to short-circuit because Loki currently handles duplicates by taking the `last seen` value.

Reply:
🙈 I'm sorry